File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

map<string, int> m;

vector<int> solution(vector<string> gems) {
    vector<int> answer(2);
    
    for(string gem: gems) {
        m[gem] = 0;
    }
    
    int left = 0, right = 0, cur_gem = 1, total_gem = m.size(), min_len = gems.size() + 1;
    m[gems[right]] = 1;
    
    while(left <= right) {
        if(cur_gem == total_gem) {
            if(right - left + 1 < min_len) {
                min_len = right - left + 1;
                answer[0] = left + 1;
                answer[1] = right + 1;
            }
            m[gems[left]]--;
            if(m[gems[left]] == 0) cur_gem--;
            left++;
        }
        else {
            right++;
            if(right >= gems.size()) break;
            if(m[gems[right]] == 0) cur_gem++;
            m[gems[right]]++;
        }
    }
    
    return answer;
}

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
from collections import defaultdict

def solution(gems):
    unique_gems = set(gems)
    cur_unique_gems = 0
    gem_count = defaultdict(int)
    end = 0
    min_len = len(gems) + 1
    answer = [1, len(gems)]
    
    for start in range(len(gems)):
        while cur_unique_gems < len(unique_gems) and end < len(gems):
            gem_count[gems[end]] += 1
            if gem_count[gems[end]] == 1:
                cur_unique_gems += 1
            end += 1

        if cur_unique_gems == len(unique_gems) and end - start + 1 < min_len:
            answer = [start + 1, end]
            min_len = end - start + 1
            
        gem_count[gems[start]] -= 1
        if gem_count[gems[start]] == 0:
            cur_unique_gems -= 1
    
    return answer